### PR TITLE
Harmonize dialogs by adding apply, cancel, and reset buttons

### LINKED ISSUE #229: Harmonize pop-ups and dialogs
Currently, pop-ups (e.g., column filtering, etc) differ with respect to their (closing) behaviour: Ok/Apply/Reset vs instant vs mouse auto-hide vs mouse click outside to hide.

Please harmonize this according to the following ideas:

* Cancel/apply should be explicit (no auto-hide)
* On apply / cancel close whole menu
* Add labels to apply / cancel
* Apply instantly in the background when possible (e.g., filtering), however go back to previous state if setting is cancelled

### PR DESCRIPTION
**Pull request by sgratzl**
closes #229, #272, #277 

**prerequisites**: 
 * [x] branch is up-to-date with the branch to be merged with, i.e. develop
 * [x] build is successful
 * [x] code is cleaned up and formatted 
 * [ ] tested with Firefox 52, Firefox 57+, Chrome 64+, MS Edge 16+

### Summary
 
unifies the dialogs as described in the issue

### CODE REVIEW

**Comment by thinkh** [comment]
Thanks for the first implementation. I ran a first quick test and it looks already very promising. I found a couple of things that needs to be addressed.

1. It looks like the live preview for filter and visualization dialogs are disabled. For the filter dialog the live preview works for the filters in the side panel, but it does not for the column filter. Can we enable the live preview for the filter dialogs (including the categorical filter dialog) and visualization by default? I think as a rule of thumb, all dialogs that had no cancel/reset/apply button before should have a live preview now, but in order to confirm the change the user must click on the apply button.

lineup-disabled-live-preview

2. When I open a sub-dialog from the more menu, change something and apply the changes, the list of more values remains open. In previous LineUp versions the dialog and the more menu were closed on apply.

lineup-close-more-menu

**Comment by sgratzl** [comment]
re 1: done

re 2: when I try https://lineup.js.org/app_develop/#soccer it behaves the same. just clicking outside will close all dialogs. A proper dialog ending (cancel, apply) won't close the more dialog

File: src/config.ts [hunk 1]
```diff
@@ -92,9 +92,9 @@ export interface ILineUpOptions {
 
   /**
    * flag whether to shows filter previews as soon as the user changes the filter in the dialog
-   * @default false
+   * @default ['filter', 'vis']
    */
-  liveFilterPreviews: boolean;
+  livePreviws: ('search' | 'filter' | 'vis')[];
```
Comment (thinkh) [inline src/config.ts]: Typo: `livePreviews`

**Comment by thinkh** [comment]
We tested the dialogs and the live preview in more detail and saw that the following dialogs had a live preview (instant feedback) before, but missing it in this implementation.

**Number Column**
* Sort By
* Group By
* Sort Groups By
* Color Mapping

**Numbers Column**
* Sort By
* Color Mapping

**String Column**
* Search (shortcut + menu)
* Sort By
* Group By -> was not available before, but would be consistent
* Filter -> was not available before, but would be consistent (see also #272)

**Date Column**
* Sort By
* Group By
* Sort Groups By

**Categorical Column**
* Sort By
* Group By
* Sort Groups By

@sgratzl Would it be possible to (re-)enable the live preview for the above mentioned dialogs?

**Comment by sgratzl** [comment]
> @sgratzl Would it be possible to (re-)enable the live preview for the above mentioned dialogs?

done

**Comment by thinkh** [comment]
Thanks, that looks very nice now! 😃 

I tested this branch in Ordino and have a minor change request. At the moment we cannot distinguish between live preview and "applied actions", i.e., also the events from the live preview are pushed to the provenance graph. Could you maybe add a `preview` flag (or similar) to the event so that we can sort these events out? At the end only the  applied actions should be pushed into the provenance graph.

lineup-ordino-live-preview-prov-graph

**Comment by thinkh** [comment]
Unfortunately I found another bug, which might be releated to my previous comment.

**Steps to reproduce**

1. Open a String column filter dialog
2. Enter nothing
3. Click outside or cancel

**Observed behavior**

The filter icon is active, even though nothing has changed. (It is also tracked in the provenance graph.) Interestingly it seems just to appear the first time(?).

lineup-string-filter


**Expected behavior**

If nothing has changed (i.e., for the string column filter no text is entered) and the user cancels the filter. No filter should be set.

**Comment by sgratzl** [comment]
> If nothing has changed (i.e., for the string column filter no text is entered) and the user cancels the filter. No filter should be set.

done

**Comment by sgratzl** [comment]
> I tested this branch in Ordino and have a minor change request. At the moment we cannot distinguish between live preview and "applied actions", i.e., also the events from the live preview are pushed to the provenance graph. Could you maybe add a `preview` flag (or similar) to the event so that we can sort these events out? At the end only the applied actions should be pushed into the provenance graph.

I don't know how to do that. From a model point of view there is no difference whether it is preview or normal set operation. Same goes for changing grouping or sorting criteria (as in your example) (which is an operation on the Ranking itself). Moreover, different events have different argument lists so there is no fixed place where to put the `preview` flag. Adding this would require to change the whole event logic

**Comment by thinkh** [comment]
Thanks for fixing the string filter bug.

I see your point, regarding the `preview` flag in the events. Alternatively, would it be possible to add further events when user (a) opens a dialog and (b) applies the changes? With that event combination we should be able to filter out the live preview events.

**Comment by thinkh** [comment]
I found a similar bug as the (recently reported) string column filter with the categorical column filter. Can you please have a look for this again?

lineup-categorical-filter

**Comment by sgratzl** [comment]
> . Alternatively, would it be possible to add further events when user (a) opens a dialog and (b) applies the changes? With that event combination we should be able to filter out the live preview events.

which instance should trigger those events? it cannot be part of any model (column, ranking, data provider) since it is purely UI related

**Comment by thinkh** [comment]
> which instance should trigger those events? it cannot be part of any model (column, ranking, data provider) since it is purely UI related

What about the LineUp instance itself? Alternatively you could add a dedicated event dispatcher for UI releated events.

File: demo/dialog.html [hunk 2]
```diff
@@ -49,6 +49,14 @@
       })
     }
     const lineup = LineUpJS.asTaggle(document.querySelector('article div'), arr);
+
+    lineup.on('dialogOpened', (dialog) => {
```
Comment (thinkh) [inline demo/dialog.html]: @sgratzl Do you have an idea how to use the constants in the JS version?
They are exported from the `ALineUp` or `DialogManager` class, but I cannot use neither of them.

```
  static readonly EVENT_DIALOG_OPENED = DialogManager.EVENT_DIALOG_OPENED;
  static readonly EVENT_DIALOG_CLOSED = DialogManager.EVENT_DIALOG_CLOSED;
```

**Comment by thinkh** [comment]
Thanks for fixing #277 and implementing the events. With this addition we should be able to filter out the live preview events. In general it works now very smooth. While testing I just found two bugs:

#### 1. Reset does not work for missing value rows in categorical column filter 

1. Filter missing values of categorical column
1. Confirm dialog
1. Open filter dialog again and click reset

Nothing happens. As I user, I would expect that the checkbox is unchecked on reset.

lineup-reset-categorical-filter

#### 2. Cannot filter missing values in numbers column

1. Open the _missing.html_ demo 
1. Open the filter dialog for the numbers column

There are no missing value rows found, even though there are clearly some in the dataset. As I user, I would expect that the checkbox works and I can filter out missing values rows.

grafik

**Comment by sgratzl** [comment]
done

**Pull request merged**